### ISSUE TITLE
Saving cloned object returns same objectId

### ISSUE THREAD

**Issue by sadakchap** [parse-community/parse-server#7422]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description

Saving a cloned object which had `authData` set, returns new object with same `objectId` for `_User` class.

### Steps to reproduce

1. Clone a User object with `authData` set
2. save that object using `Parse.Object.saveAll`
3. or try this code

```js
let q = new Parse.Query('_User');
q.equalTo('objectId', 'EiWF4cUw9P'); // any User objectId which has authData set
const res = await q.find({ useMasterKey: true });
let toClone=[];
for (const object of res) {
  toClone.push(object.clone());
}
await Parse.Object.saveAll(toClone, { useMasterKey: true });
```

### Actual Outcome

returns object with same `objectId` & other fields as well.

### Expected Outcome

should have new `objectId` & unique `username`

### Failing Test Case / Pull Request

- [ ] 🤩 I submitted a PR with a fix and a test case.
- [ ] 🧐 I submitted a PR with a failing test case.

###  Environment

Server
- Parse Server version: `4.5.0`
- Operating system: `Ubuntu`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `local`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `4.4`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `irrelevant`
- SDK version: `irrelevant`

**Comment by mtrezza** [comment]
Thanks for reporting.

Does this only occur with `saveAll` or also if you `save` a cloned object that contains authData?

**Comment by sadakchap** [comment]
Both `saveAll` and `save`.

**Comment by mtrezza** [comment]
Good to know! Could you write a failing test and submit it as a PR? The test should be simple as possible, using `save`.

**Comment by sadakchap** [comment]
Sure @mtrezza .

I think, internally `save` is calling this [RestWrite.js](https://github.com/parse-community/parse-server/blob/master/src/RestWrite.js#L1) which clearly states https://github.com/parse-community/parse-server/blob/d8dc524c08036b2f326730e1ed214ed7a76141cb/src/RestWrite.js#L3

And in above code snippet, when we are cloning the object, it also clones `authData` and sends same authData to `save`, which updates the object instead of creating new object. Hence returning same `objectId` which I think is expected behaviour.

**Comment by mtrezza** [comment]
> Is this expected behaviour ? If yes, then can we remove authData while cloning a User object.

I would say an object should solely be identified by its ID, not by a field that is unique and only _infers_ the object ID. Inferring is opaque and seems like it could become a potential exploit, depending on future code changes.

I think we should analyze this behavior in a separate issue and possibly remove that inference.

For closing it definitely doesn't male sense to duplicate authData, as it doesn't with duplicating the password or any other unique field. So that can be removed.

**Comment by sadakchap** [comment]
> I would say an object should solely be identified by its ID, not by a field that is unique and only _infers_ the object ID. Inferring is opaque and seems like it could become a potential exploit, depending on future code changes.
> 
> I think we should analyze this behavior in a separate issue and possibly remove that inference.

Yes, sounds good. But, could you please also confirm this behaviour. I am not very sure as I am not very familiar with this code. :pleading_face: 

> For cloning it definitely doesn't make sense to duplicate `authData`, as it doesn't with duplicating the username and password, or any other unique field. So `authData` can be removed.

ok, should I update this [pr](https://github.com/parse-community/parse-dashboard/pull/1697) for not cloning `authData` or should it be in [Parse-SDK-JS](https://github.com/parse-community/Parse-SDK-JS) repo . 

If in Parse-SDK-JS repo , then could you please tell me where are we specifying `clone` not to copy `username` or unique fields. I tried it, but I could not find it. [This](https://github.com/parse-community/Parse-SDK-JS/blob/005321fa9d820b7f5d5275c6c77b842c449b44eb/src/ParseObject.js#L936) is where I am looking.

Or it is that I am looking in wrong file ?

Or if should I update this https://github.com/parse-community/parse-dashboard/pull/1697, then it simply unsetting the `authData` attribute.. perhaps the easiest approach now.

Please let me know about your thoughts.

**Comment by mtrezza** [comment]
The update would be in the JS SDK. Otherwise cloning in dashboard would be different than cloning programmatically.

The cloning method is [here](https://github.com/parse-community/Parse-SDK-JS/blob/ffc523f9cd0fa2dd3fcc9754d5f15c3c1c78314f/src/ParseObject.js#L936-L955).
As you can see in the [condition](https://github.com/parse-community/Parse-SDK-JS/blob/ffc523f9cd0fa2dd3fcc9754d5f15c3c1c78314f/src/ParseObject.js#L945), all fields are cloned except for read only properties. Read-only properties are for example `createdAt` and `objectId`.

There is a [cloning test](https://github.com/parse-community/Parse-SDK-JS/blob/e6f15fdc09429a2c456ec6387d365914af8a8988/src/__tests__/ParseUser-test.js#L117-L130) that shows that the username is cloned as well. I wonder how this works in the dashboard when cloning a user, if the clone has the same username. Do you currently manually remove the username in the dashboard logic?

**Comment by sadakchap** [comment]
> I am not sure whether the change should be made in the JS SDK or in dashboard. We need to understand why it is that cloning a user object with `authData` leads to updating the user object. This seems strange and should be investigated. If that is a bug, then the bug fix will solve this. If it is not a bug but indeed desired behavior, then `authData` can simply be deleted in the dashboard logic.

[This](https://github.com/parse-community/parse-server/blob/d8dc524c08036b2f326730e1ed214ed7a76141cb/src/RestWrite.js#L473-L567) method is responsible for handling creation or updation of User objects with `authData`. I think it first tries to find objects with same `authData` and if there is 1 record, it tries to update it. But, I could not the understand the creation path for objects with `authData`.


> I wonder how this works in the dashboard when cloning a user, if the clone has the same username, but username is unique. Do you currently manually remove the username in the dashboard logic?

Since, dashboard never has `password` field set, we try to save cloned User object without `password` and request fails with `201` error. And we are checking this error code & showing option to edit the cloned rows to user. Please see this https://github.com/parse-community/parse-dashboard/pull/1697/commits/53cd6e6c24d9365fdab70a700480ddb5214f4e1d

**Comment by mtrezza** [comment]
What about the username which is transmitted to the dashboard - are you removing this in dashboard code?

**Comment by sadakchap** [comment]
> What about the username which is transmitted to the dashboard - are you removing this in dashboard code?

No. I am not removing `username` from dashboard.

**Comment by mtrezza** [comment]
How is the object then cloned in dashboard, if the cloned user has the same `username`?

**Comment by sadakchap** [comment]
Cloning fails, as server validation fails cause no `password` field was provided.

**Comment by mtrezza** [comment]
Yes, but the failed row then does not display a username. That is why I'm asking, are you removing the username? You said no, but why doesn't it appear in the field then?

**Comment by sadakchap** [comment]
After this https://github.com/parse-community/parse-dashboard/pull/1697/commits/4cb326dcfaf6dc7693075aae81df2d9a72cc737a commit, I am explicitly removing `username` & `authData`. I also mentioned the in https://github.com/parse-community/parse-dashboard/pull/1697#issuecomment-868220368

But before this [commit](https://github.com/parse-community/parse-dashboard/commit/4cb326dcfaf6dc7693075aae81df2d9a72cc737a), I was not removing the username, its value was there in failed clone row. You can also try in commits prior this [commit](https://github.com/parse-community/parse-dashboard/commit/4cb326dcfaf6dc7693075aae81df2d9a72cc737a).
I removed username & authData in https://github.com/parse-community/parse-dashboard/pull/1697/commits/4cb326dcfaf6dc7693075aae81df2d9a72cc737a , because as we discussed that username & authData does not make sense to be duplicated and cloning a user object duplicates them, so I thought of removing them.
